cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -109,13 +112,100 @@ func (spm *KubeStaticPodPathManager) BackupManifestDir() string {
+	backupManifestPath := pathMgr.BackupManifestPath(component)
+
+	// Store the backup path in the recover list. If something goes wrong now, this component will be rolled back.
+	recoverManifests[component] = backupManifestPath
+
+	// Move the old manifest into the old-manifests directory
+	if err := pathMgr.MoveFile(currentManifestPath, backupManifestPath); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	// Move the new manifest into the manifests directory
+	if err := pathMgr.MoveFile(newManifestPath, currentManifestPath); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Moved upgraded manifest to %q and backed up old manifest to %q\n", currentManifestPath, backupManifestPath)
+	fmt.Println("[upgrade/staticpods] Waiting for the kubelet to restart the component")
+
+	// Wait for the mirror Pod hash to change; otherwise we'll run into race conditions here when the kubelet hasn't had time to
+	// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+	// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+	// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
+	if err := waiter.WaitForStaticPodControlPlaneHashChange(cfg.NodeName, component, beforePodHash); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	// Wait for the static pod component to come up and register itself as a mirror pod
+	if err := waiter.WaitForPodsWithLabel("component=" + component); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Component %q upgraded successfully!\n", component)
+	return nil
+}
 
-	// This string-string map stores the component name and backup filepath (if a rollback is needed).
-	// If a rollback is needed,
+func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, recoverManifests map[string]string) error {
+	// Add etcd static pod spec only if external etcd is not configured
+	if len(cfg.Etcd.Endpoints) != 0 {
+		return fmt.Errorf("external etcd cannot be upgraded with kubeadm")
+	}
+
+	// Need to check currently used version and version from constants, if differs then uprgade
+	runningEtcdVersion, err := util.GetEtcdRunningVersion()
+	if err != nil {
+		return err
+	}
+	targetKubernetesVersion := version.MustParseGeneric(cfg.KubernetesVersion)
+	upgradeEtcdVersion, err := constants.GetEtcdSupportedVersion(uint8(targetKubernetesVersion.Minor()))
+	if err != nil {
+		return err
+	}
+	newEtcdVersion := version.MustParseGeneric(upgradeEtcdVersion)
+	if err != nil {
+		return err
+	}
+	// Ignore error here because the version struct was built with MustParse method
```

Comment:
don't we need to update `cfg.Etcd.Image` somewhere to include the new version? Otherwise it will continue using the [default version](https://github.com/kubernetes/kubernetes/blob/3e245366f0ead632c57e406e3d17987a375cb4e0/cmd/kubeadm/app/images/images.go#L34)

Reply:
done